nautobot/extras/models/groups.py

```diff
@@ -313,13 +316,78 @@ def get_queryset(self):
 
     @property
     def members(self):
-        """Return the member objects for this group."""
+        """Return the member objects for this group, never cached."""
         # If there are child groups, return the generated group queryset, otherwise use this group's
         # `filter` directly.
         if self.children.exists():
             return self.get_group_queryset()
         return self.get_queryset()
 
+    @property
+    def members_cached(self):
+        """Return the member objects for this group, cached if available."""
+        cache_key = f"{self.__class__.__name__}.{self.id}.cached_members"
+
+        unpickled_query = None
+        try:
+            cached_query = cache.get(cache_key)
+            if cached_query is not None:
+                unpickled_query = pickle.loads(cached_query)
+        except pickle.UnpicklingError:
+            logger.warning("Failed to unpickle cached members for %s", self)
+        finally:
+            if unpickled_query is None:
+                unpickled_query = self.members.all()
+                cached_query = pickle.dumps(unpickled_query)  # Explicitly pickle the query to evaluate it.
+                cache.set(cache_key, cached_query, get_settings_or_config("DYNAMIC_GROUPS_MEMBER_CACHE_TIMEOUT"))
+
+        return unpickled_query
+
+    def get_members(self, skip_cache=False, force_update_cache=False):
+        """
+        Return the member objects for this group.
+
+        Args:
+            skip_cache (bool): Whether to skip the cache and run the query directly. Optional. Default: False.
+            force_update_cache (bool): Whether to force an update of the cache. Optional. Default: False. Ignored if `skip_cache` is True.
+        """
+
+        if skip_cache:
+            return self.members
+
+        if force_update_cache:
+            cache_key = f"{self.__class__.__name__}.{self.id}.cached_members"
+            cache.delete(cache_key)
+
+        return self.members_cached
+
+    def am_i_a_member(self, obj, skip_cache=False):
+        """
+        Return True if the given object is a member of this group.
+
+        Does check if object's content type matches this group's content type.
+
+        Args:
+            obj (django.db.models.Model): The object to check for membership.
+            skip_cache (bool): Whether to skip the cache and run the query directly. Optional. Default: False.
+
+        Returns:
+            bool: True if the object is a member of this group, otherwise False.
+        """
+
+        # Object's class may have content type cached, so check that first.
+        if not skip_cache and hasattr(type(obj), "_content_type"):
```

Comment:
I feel like we *ought* to be able to assert that any object we're checking for DG membership *must* be a `BaseModel` subclass and therefore should always have the `_content_type` property?

Reply:
So following the style rulings: https://stackoverflow.com/questions/21024913/python-isinstance-vs-hasattr-vs-try-except-what-is-better

The preference (it seems) is if you can assume something should work, it's better to do it and handle the failure (save the computational check I guess).

Also, it seems torn on the `isinstance` vs `hasatter`; some seem to suggest duck typing is better than inheritance checking.

Can we get an internal style guide on this?